gno.land/cmd/gnofaucet/throttle.go

```diff
@@ -1,60 +1,92 @@
 package main
 
 import (
-	"net"
+	"context"
+	"errors"
+	"net/netip"
+	"sync"
 	"time"
+)
 
-	"github.com/gnolang/gno/tm2/pkg/service"
+const (
+	maxRequestsPerIP = uint64(5)
 )
 
-type SubnetThrottler struct {
-	service.BaseService
-	ticker   *time.Ticker
-	subnets3 [2 << (8 * 3)]uint8
-	// subnets2 [2 << (8 * 2)]uint8
-	// subnets1 [2 << (8 * 1)]uint8
+var errInvalidNumberOfRequests = errors.New("invalid number of requests")
+
+type requestMap map[netip.Addr]uint64
+
+// iterate ranges over the request map
+func (r requestMap) iterate(cb func(key netip.Addr, value uint64)) {
+	for ip, requests := range r {
+		cb(ip, requests)
+	}
 }
 
-func NewSubnetThrottler() *SubnetThrottler {
-	st := &SubnetThrottler{}
-	// st.ticker = time.NewTicker(time.Second)
-	st.ticker = time.NewTicker(time.Minute)
-	st.BaseService = *service.NewBaseService(nil, "SubnetThrottler", st)
-	return st
+type ipThrottler struct {
+	throttleTimeout time.Duration
+	requestMap      requestMap
+
+	sync.Mutex
 }
```

Comment:
why not golang.org/x/time/rate ?:

```go
rl := rate.NewLimiter(rate.Every(10*time.Second), 50) // 50 request every 10 seconds
[...]
err := rl.Wait(ctx) // This is a blocking call. Honors the rate limit
if err != nil {
	return nil, err
}
```

Reply:
I've applied this suggestion here:

[7349c9b](https://github.com/gnolang/gno/pull/1614/commits/7349c9ba626d2210998aae09234efe755067b1df)